src/transformations/execute.ts

```diff
@@ -161,38 +162,64 @@ export async function executeTransformations(
 
       case 'macro':
         const recordedMacro = (await Register.get(transformation.register))?.text;
-        if (!(recordedMacro instanceof RecordedState)) {
+        if (!recordedMacro) {
           return;
-        }
 
-        vimState.isReplayingMacro = true;
+        } else if (typeof recordedMacro === 'string') {
+          // A string was set to the register. We need to execute the characters as if they were typed (in normal mode).
+          const keystrokes = keystrokesExpressionParser.parse(recordedMacro);
+          if (!keystrokes.status) {
+            return;
+          }
+
+          vimState.isReplayingMacro = true;
+
+          vimState.recordedState = new RecordedState();
+          await modeHandler.handleMultipleKeyEvents(keystrokes.value);
+
+          // Set the executed register as the registerName, otherwise the last action register is used.
+          vimState.recordedState.registerName = transformation.register;
+
+          vimState.lastInvokedMacro = vimState.recordedState;
+          vimState.isReplayingMacro = false;
+
+          if (vimState.lastMovementFailed) {
+            // movement in last invoked macro failed then we should stop all following repeating macros.
+            // Besides, we should reset `lastMovementFailed`.
+            vimState.lastMovementFailed = false;
+            return;
+          }
 
-        vimState.recordedState = new RecordedState();
-        if (transformation.register === ':') {
-          await new ExCommandLine(recordedMacro.commandString, vimState.currentMode).run(vimState);
-        } else if (transformation.replay === 'contentChange') {
-          await modeHandler.runMacro(recordedMacro);
         } else {
```

Comment:
Any reason you see that we can't (eventually) drop this `else` entirely and always store registers as text, rather than `RecordedState`s?

Reply:
The only reason I can think of is that using `RecordedState` might have better runtime performance.

Reply:
Yeah, we'll need to measure that